helix-term/src/application.rs

```diff
@@ -747,31 +748,63 @@ impl Application {
+                                if let Some(version) = params.version {
+                                    if version != doc.version() {
+                                        log::info!("Version ({version}) is out of date for {path:?} (expected ({}), dropping PublishDiagnostic notification", doc.version());
+                                        return false;
+                                    }
                                 }
-                            }
-
-                            true
-                        });
+                                true
+                            });
 
                         if let Some(doc) = doc {
-                            let lang_conf = doc.language_config();
-                            let text = doc.text();
+                            let lang_conf = doc.language.clone();
+                            let text = doc.text().clone();
+
+                            let mut unchaged_diag_sources_ = Vec::new();
+                            if let Some(lang_conf) = &lang_conf {
+                                if let Some(old_diagnostics) =
+                                    self.editor.diagnostics.get(&params.uri)
+                                {
+                                    if !lang_conf.persistent_diagnostic_sources.is_empty() {
+                                        // Sort diagnostics first by severity and then by line numbers.
+                                        // Note: The `lsp::DiagnosticSeverity` enum is already defined in decreasing order
+                                        params
+                                            .diagnostics
+                                            .sort_unstable_by_key(|d| (d.severity, d.range.start));
+                                    }
+                                    for source in &lang_conf.persistent_diagnostic_sources {
+                                        let new_diagnostics = params
+                                            .diagnostics
+                                            .iter()
+                                            .filter(|d| d.source.as_ref() == Some(source));
+                                        let old_diagnostics = old_diagnostics
+                                            .iter()
+                                            .filter(|(d, d_server)| {
+                                                *d_server == server_id
+                                                    && d.source.as_ref() == Some(source)
+                                            })
+                                            .map(|(d, _)| d);
+                                        if new_diagnostics.eq(old_diagnostics) {
+                                            unchaged_diag_sources_.push(source.clone())
+                                        }
+                                    }
+                                }
+                            }
```

Comment:
I wonder if all this logic could just move into `replace_diagnostics`. As it is, the late function has been changed so that it assumes repeat diagnostics have already been removed, and could potentially allow repeats to be added if this invariant is violated.

Reply:
that's not really possible because we need access to `Editor::diagnostics` (not the same as `Document::diagnostics`)  which would be pretty awkward. I also actually prefer to keep the LSP specific logic like thous out of the document if possible.

Reply:
that said this call site doesn't actually handle this correctly. i think I messed that up in the last rebase by accident. I need to rebase this again anyway so I will take a look at that then